### PR TITLE
🌱 test: improve autoscale tests for to/from zero and running autoscaler in bootstrap cluster

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Adjusts autoscale tests to:

* Allow running cluster-autoscaler inside the management cluster
  * Note: for providers there may be no connectivity from the workload to the management cluster running in kind
* make machinepool's optional for the test:
  * Note: not all providers support machinepools (e.g. CAPV)
* add option for scale to/from zero tests

/area e2e-testing

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: test/framework/autoscaler_helpers.go [hunk 1]
```diff
@@ -143,7 +157,7 @@ type AddScaleUpDeploymentAndWaitInput struct {
 func AddScaleUpDeploymentAndWait(ctx context.Context, input AddScaleUpDeploymentAndWaitInput, intervals ...interface{}) {
 	By("Create a scale up deployment with resource requests to force scale up")
 	if input.ContainerImage == "" {
-		input.ContainerImage = "registry.k8s.io/pause"
+		input.ContainerImage = "registry.k8s.io/pause:3.10"
```
Comment (chrischdi) [inline test/framework/autoscaler_helpers.go:160]: Let's not reference latest, did e.g. not work on arm64.

File: test/framework/deployment_helpers.go [hunk 2]
```diff
@@ -479,7 +479,7 @@ func DeployUnevictablePod(ctx context.Context, input DeployUnevictablePodInput)
 					Containers: []corev1.Container{
 						{
 							Name:  "web",
-							Image: "registry.k8s.io/pause:latest",
+							Image: "registry.k8s.io/pause:3.10",
```
Comment (chrischdi) [inline test/framework/deployment_helpers.go:482]: Let's not reference latest, did e.g. not work on arm64.

File: test/framework/machinedeployment_helpers.go [hunk 3]
```diff
@@ -648,5 +648,6 @@ func AssertMachineDeploymentReplicas(ctx context.Context, input AssertMachineDep
 		g.Expect(input.Getter.Get(ctx, key, md)).To(Succeed(), fmt.Sprintf("failed to get MachineDeployment %s", klog.KObj(input.MachineDeployment)))
 		g.Expect(md.Spec.Replicas).Should(Not(BeNil()), fmt.Sprintf("MachineDeployment %s replicas should not be nil", klog.KObj(md)))
 		g.Expect(*md.Spec.Replicas).Should(Equal(input.Replicas), fmt.Sprintf("MachineDeployment %s replicas should match expected replicas", klog.KObj(md)))
+		g.Expect(md.Status.Replicas).Should(Equal(input.Replicas), fmt.Sprintf("MachineDeployment %s status.replicas should match expected replicas", klog.KObj(md)))
```
Comment (chrischdi) [inline test/framework/machinedeployment_helpers.go:651]: This ensures the rollout finished and is stable.

File: test/framework/machinepool_helpers.go [hunk 4]
```diff
@@ -365,5 +365,6 @@ func AssertMachinePoolReplicas(ctx context.Context, input AssertMachinePoolRepli
 		g.Expect(input.Getter.Get(ctx, key, mp)).To(Succeed(), fmt.Sprintf("failed to get MachinePool %s", klog.KObj(input.MachinePool)))
 		g.Expect(mp.Spec.Replicas).Should(Not(BeNil()), fmt.Sprintf("MachinePool %s replicas should not be nil", klog.KObj(mp)))
 		g.Expect(*mp.Spec.Replicas).Should(Equal(input.Replicas), fmt.Sprintf("MachinePool %s replicas should match expected replicas", klog.KObj(mp)))
+		g.Expect(mp.Status.Replicas).Should(Equal(input.Replicas), fmt.Sprintf("MachinePool %s replicas should match expected replicas", klog.KObj(mp)))
```
Comment (chrischdi) [inline test/framework/machinepool_helpers.go:368]: This ensures the rollout finished and is stable. Otherwise the expected numbers in the next scale steps may be wrong because the scale up deployment considers nodes which will vanish.

**Comment by chrischdi** [comment]
Can't explain it. Running in parallel again and testing locally

/retest

File: test/e2e/autoscaler.go [hunk 5]
```diff
@@ -145,21 +156,27 @@ func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput)
...
 		// Get a ClusterProxy so we can interact with the workload cluster
 		workloadClusterProxy := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, clusterResources.Cluster.Namespace, clusterResources.Cluster.Name)
 		mdOriginalReplicas := *clusterResources.MachineDeployments[0].Spec.Replicas
 		Expect(strconv.Itoa(int(mdOriginalReplicas))).To(Equal(mdNodeGroupMinSize), "MachineDeployment should have replicas as defined in %s", clusterv1.AutoscalerMinSizeAnnotation)
-		mpOriginalReplicas := *clusterResources.MachinePools[0].Spec.Replicas
-		Expect(int(mpOriginalReplicas)).To(Equal(1), "MachinePool should default to 1 replica via the MachinePool webhook")
+
+		var mpOriginalReplicas int32
+		if hasMachinePool {
+			mpOriginalReplicas := *clusterResources.MachinePools[0].Spec.Replicas
```
Comment (sbueringer) [inline test/e2e/autoscaler.go:177]: This creates a local scoped mpOriginalReplicas variable, is this intentional?

File: test/e2e/autoscaler.go [hunk 6]
```diff
@@ -226,6 +245,46 @@ func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput)
...
+				Cluster:                     clusterResources.Cluster,
+				NodeGroupMinSize:            "0",
+				NodeGroupMaxSize:            mdNodeGroupMaxSize,
+				WaitForAnnotationsToBeAdded: input.E2EConfig.GetIntervals(specName, "wait-autoscaler"),
+			})
+
+			By("Scaling the MachineDeployment scale up deployment to zero")
+			framework.ScaleScaleUpDeploymentAndWait(ctx, framework.ScaleScaleUpDeploymentAndWaitInput{
+				ClusterProxy: workloadClusterProxy,
+				Replicas:     mpOriginalReplicas + 0,
```
Comment (sbueringer) [inline test/e2e/autoscaler.go:262]: What does this have to do with MachinePools?

(also doesn't look like it's always 0)

File: test/e2e/autoscaler.go [hunk 7]
```diff
@@ -226,6 +245,46 @@ func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput)
...
+				Getter:                   input.BootstrapClusterProxy.GetClient(),
+				MachineDeployment:        clusterResources.MachineDeployments[0],
+				Replicas:                 0,
+				WaitForMachineDeployment: input.E2EConfig.GetIntervals(specName, "wait-controllers"),
+			})
+
+			By("Scaling the MachineDeployment scale up deployment to 1")
+			framework.ScaleScaleUpDeploymentAndWait(ctx, framework.ScaleScaleUpDeploymentAndWaitInput{
+				ClusterProxy: workloadClusterProxy,
+				Replicas:     mpOriginalReplicas + 1,
```
Comment (sbueringer) [inline test/e2e/autoscaler.go:276]: Same here, is this always 1, why MP?

File: test/e2e/autoscaler.go [hunk 8]
```diff
@@ -239,67 +298,127 @@ func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput)
...
+			// Since we scaled up the MachinePool manually and the workload has not changed auto scaler
+			// should detect that there are unneeded nodes and scale down the MachinePool.
+			framework.AssertMachinePoolReplicas(ctx, framework.AssertMachinePoolReplicasInput{
+				Getter:             input.BootstrapClusterProxy.GetClient(),
+				MachinePool:        clusterResources.MachinePools[0],
+				Replicas:           mpScaledUpReplicas,
+				WaitForMachinePool: input.E2EConfig.GetIntervals(specName, "wait-controllers"),
+			})
+
+			if input.ScaleToAndFromZero {
```
Comment (sbueringer) [inline test/e2e/autoscaler.go:364]: Is this feature specified for MachinePools? (they don't have an InfraMachineTemplate)

Mabye let's focus on MD for now and add a disclaime to the ScaleToAndFromZero input field?

File: test/e2e/autoscaler.go (hunk 5 above)
Comment (chrischdi) [reply test/e2e/autoscaler.go]: nope, that was the bug why the test failed :-)

File: test/e2e/autoscaler.go [hunk 9]
```diff
@@ -239,67 +298,127 @@ func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput)
...
+				By("Checking the MachineDeployment finished scaling up")
+				framework.AssertMachinePoolReplicas(ctx, framework.AssertMachinePoolReplicasInput{
+					Getter:             input.BootstrapClusterProxy.GetClient(),
+					MachinePool:        clusterResources.MachinePools[0],
+					Replicas:           1,
+					WaitForMachinePool: input.E2EConfig.GetIntervals(specName, "wait-controllers"),
+				})
+			}
+
+			By("Disabling the autoscaler for MachineDeployments to test MachinePools")
```
Comment (sbueringer) [inline test/e2e/autoscaler.go:409]: Wondering why we do this at this point

Comment (chrischdi) [reply test/e2e/autoscaler.go:409]: missed fixing this string :-)

These last two steps are not really necessary 👍  I'll remove them

File: test/e2e/autoscaler.go (hunk 7 above)
Comment (chrischdi) [reply test/e2e/autoscaler.go:276]: For CAPV this will always be 1.

If a provider supports both MP and MD, we need to add the number of existing machine pool nodes because the deployment can be scheduled otherwise and won't trigger a scale-up.

File: test/e2e/autoscaler.go (hunk 6 above)
Comment (chrischdi) [reply test/e2e/autoscaler.go:262]: We want to scale down the machinedeployment to 0 which means we need to set the deployment to the number of machinepool replicas (which could be 0 if machinepools are not there).

File: test/e2e/autoscaler.go (hunk 8 above)
Comment (chrischdi) [reply test/e2e/autoscaler.go:364]: 👍

File: test/e2e/autoscaler.go (hunk 7 above)
Comment (sbueringer) [reply test/e2e/autoscaler.go:277]: Thx! Let's add a godoc comment

File: test/e2e/autoscaler.go (hunk 6 above)
Comment (sbueringer) [reply test/e2e/autoscaler.go:263]: Makes sense, let's add a comment

File: test/framework/autoscaler_helpers.go [hunk 10]
```diff
@@ -247,6 +261,34 @@ func DeleteScaleUpDeploymentAndWait(ctx context.Context, input DeleteScaleUpDepl
...
 }
 
+// ScaleScaleUpDeploymentAndWaitInput is the input for ScaleScaleUpDeploymentAndWait.
+type ScaleScaleUpDeploymentAndWaitInput struct {
+	ClusterProxy ClusterProxy
+	Name         string
+	Replicas     int32
+}
+
+// ScaleScaleUpDeploymentAndWait deletes the scale up deployment and waits for it to be deleted.
```
Comment (sbueringer) [inline test/framework/autoscaler_helpers.go:271]: godoc comment seeems outdated (deletes)

File: test/e2e/autoscaler.go (hunk 8 above)
Comment (chrischdi) [reply test/e2e/autoscaler.go]: hmm according to the doc, it should work if the annotations are set.

Also for machinepools, if the inframachinepool provides `.status.capacity` (e.g. DockerMachinePool) it should work too.

File: test/e2e/autoscaler.go (hunk 7 above)
Comment (chrischdi) [reply test/e2e/autoscaler.go:277]: From the KEP for scale from zero:

> Support for MachinePools, either with the cluster autoscaler or using infrastructure provider native implementations (eg AWS AutoScalingGroups).

[[0]](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210310-opt-in-autoscaling-from-zero.md#machineset-and-machinedeployment-annotations)

Comment (chrischdi) [reply test/e2e/autoscaler.go:277]: Removed it

Comment (chrischdi) [reply test/e2e/autoscaler.go:277]: Yeah wanted to paste it at the other

File: test/e2e/autoscaler.go (hunk 8 above)
Comment (chrischdi) [reply test/e2e/autoscaler.go]: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210310-opt-in-autoscaling-from-zero.md#machineset-and-machinedeployment-annotations

Due to that I removed it

**Comment by sbueringer** [comment]
Thank you!

/lgtm
/approve
